taichi/backends/opengl/opengl_api.cpp

```diff
@@ -379,6 +391,7 @@ struct CompiledProgram::Impl {
     }
     for (const auto &ker : kernels) {
       auto guard = launcher->create_launch_guard(iov);
+      if (ker->rse.has_value()) ker->num_groups = (*ker->rse)((const char *)gtmp_base);
```

Comment:
Interesting. I guess this means in OpenGL, when the previous kernel is done, its write result is automatically synced back to `gtmp_base`?

Reply:
Yes, actually I forced it to write back, to keep data preserved during different kernel launches, see:
https://github.com/taichi-dev/taichi/blob/cee79d9f043bafec4593d21b182b2727b3f26d45/taichi/backends/opengl/opengl_api.cpp#L431-L444

NVIDIA's OpenGL doesn't keep data from previous kernel, while the Mesa OpenGL does...
Strange, right? You assigned a value to `gtmp[0]` in first kernel, then you obtain `gtmp[0]` become zero again in the next kernel...
So, despite it harms performance, we have to do it to keep compatiable with NV's bad OpenGL implementation...

Reply:
> You assigned a value to gtmp[0] in first kernel, then you obtain gtmp[0] become zero again in the next kernel...

hmm.. You probably already know this, but in CUDA, you need to explicitly call `cudaDeviceSynchronize()` to sync back the data to the host, because kernel launching is asynchronous. Metal also works the same way. But I guess OpenGL compute shader doesn't work this way? Or does Mesa's OpenGL automatically syncs under the hood?

Reply:
> to sync back the data to the host,

CUDA also does sync each launch, but just zero-copy?
I though after each kernel launch, GPU memory was preserved and no need to sync back to host?

> `cudaDeviceSynchronize()`

Maybe we should insert `glMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT);` between kernel launches...?

Reply:
Fixed by inserting `glMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT)`, seems it's the source of problem, many thanks!

Reply:
> CUDA also does sync each launch, but just zero-copy?

Not really. You would need to explicitly call `cudaDeviceSynchronize`, which is why Taichi has this

https://github.com/taichi-dev/taichi/blob/5fd30e61cbe482483925cb73c5c3b2109540d76f/taichi/program/program.cpp#L318


> I though after each kernel launch, GPU memory was preserved and no need to sync back to host?

OK., I think there are two separate issues here:

1. Whether the GPU memory is preserved after a launch
2. Whether the GPU memory is synced back to host memory.

`cudaDeviceSynchronize` talks about 2. For problem 1, I remember that for the *rendering* pipeline, some texture writes in the fragment shader won't be preserved, in order to improve performance. (Whether the texture write is preserved can be configured on Metal, I'm sure OpenGL has similar concepts?)

> Fixed by inserting glMemoryBarrier(GL_SHADER_STORAGE_BARRIER_BIT)

Cool!